src/Common/tests/Collections/IListTest.cs

```diff
@@ -0,0 +1,1692 @@
+                    list.Add(null);
+                    items[16] = null;
+                }
+
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddDuplicateValue()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                object[] items;
+                var list = GetList(null);
+                if (ItemsMustBeUnique)
+                {
+                    items = GenerateItems(16);
+                    list.AddRange(items);
+                    Assert.Throws<ArgumentException>(
+                        () => list.Add(items[0]));
+                }
+                else
+                {
+                    items = new object[32];
+                    var tempItems = GenerateItems(16);
+                    list.AddRange(tempItems);
+                    list.AddRange(tempItems);
+
+                    Array.Copy(tempItems, 0, items, 0, 16);
+                    Array.Copy(tempItems, 0, items, 16, 16);
+                }
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddAndClear()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                var list = GetList(null);
+                list.AddRange(GenerateItems(16));
+                list.Clear();
+                var items = GenerateItems(16);
+                list.AddRange(items);
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddAndRemove()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                var items = new object[23];
+                var list = GetList(null);
+                var tempItems = GenerateItems(32);
+                for (int i = 0; i < 16; i++)
+                {
+                    list.Add(tempItems[i]);
+                }
+                for (int i = 0; i < 16; i++)
+                {
+                    if ((i & 1) == 0)
+                    {
+                        list.RemoveAt(i/2);
+                    }
+                    else
+                    {
+                        items[i/2] = tempItems[i];
+                    }
+                }
+
+                list.RemoveAt(list.Count - 1);
+                for (int i = 0; i < 16; i++)
+                {
+                    list.Add(tempItems[16 + i]);
+                }
+
+                Array.Copy(tempItems, 16, items, 7, 16);
+
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddAndRemoveAll()
+        {
+            var list = GetList(null);
+            var tempItems = GenerateItems(16);
+            list.AddRange(tempItems);
+            foreach (object item in tempItems)
+            {
+                list.Remove(item);
+            }
+
```

Comment:
Assert.Equals(0, list.Count)

Might even want to assert that the count goes down by one each Remove.